Model/Service/Update/QueueProcessorService.php

```diff
@@ -0,0 +1,297 @@
+                    ProductUpdateQueueInterface::ACTION_VALUE_UPSERT => [],
+                    ProductUpdateQueueInterface::ACTION_VALUE_DELETE => []
+                ];
+            }
+            $totalCount += $queueEntry->getProductIdCount();
+            foreach ($queueEntry->getProductIds() as $productId) {
+                $merged[$queueEntry->getStoreId()][$queueEntry->getAction()][$productId] = $productId;
+            }
+        }
+        $mergedCount = 0;
+        foreach ($merged as $storeId => $arr) {
+            foreach ($arr as $method => $ids) {
+                // phpcs:ignore
+                $mergedCount += count($ids);
+            }
+        }
+        $this->logDebugWithStore(
+            sprintf(
+                'Merged total of %d product ids into %d',
+                $totalCount,
+                $mergedCount
+            ),
+            $store
+        );
+        return $merged;
+    }
+
+    /**
+     * Sets the timestamp for started at & updates the status to be processing
+     *
+     * @param QueueCollection $collection
+     */
+    private function setStatusToProcessing(QueueCollection $collection)
+    {
+        /* @var ProductUpdateQueueInterface $queueEntry */
+        foreach ($collection as $queueEntry) {
+            $queueEntry->setStartedAt($this->magentoTimeZone->date());
+            $queueEntry->setStatus(ProductUpdateQueueInterface::STATUS_VALUE_PROCESSING);
+        }
+    }
+
+    /**
+     * Sets the timestamp for completed at & updates the status to be done
+     *
+     * @param QueueCollection $collection
+     */
+    private function setStatusToDone(QueueCollection $collection)
+    {
+        /* @var ProductUpdateQueueInterface $queueEntry */
+        foreach ($collection as $queueEntry) {
+            $queueEntry->setCompletedAt($this->magentoTimeZone->date());
+            $queueEntry->setStatus(ProductUpdateQueueInterface::STATUS_VALUE_DONE);
+            try {
+                // phpcs:ignore
+                $this->queueRepository->save($queueEntry);
+            } catch (AlreadyExistsException $e) {
+                $this->getLogger()->exception($e);
+            }
+        }
+    }
+
+    /**
+     * Cleans up completed entries from the queue table
+     * @param Store $store
+     */
+    private function cleanupUpdateQueue(Store $store)
+    {
+        try {
+            $processed = $this->queueCollectionBuilder
+                ->init()
+                ->withCompletedHrsAgo($this->cleanupInterval)
+                ->build();
+            $this->logDebugWithStore(
+                sprintf(
+                    'Cleaning up %d entries from update queue completed %d < hours ago',
+                    $processed->count(),
+                    $this->cleanupInterval
+                ),
+                $store
+            );
+            foreach ($processed as $queueItem) {
+                // phpcs:ignore
+                $this->queueRepository->delete($queueItem);
```

Comment:
Looking at Magento code, this looks quite DB heavy to delete one by one. Perhaps we should handle all those at once in our own code?

Reply:
I agree but in this context there will not be more than a handful (something around 10) deletions per run so it should not be a biggie

Reply:
True that